src/nupic/algorithms/SpatialPooler.cpp

```diff
@@ -1120,20 +1122,86 @@ void SpatialPooler::updateDutyCyclesHelper_(vector<Real>& dutyCycles,
 
 void SpatialPooler::updateBoostFactors_()
 {
-  for (UInt i = 0; i < numColumns_; i++)
+  if (maxBoost_ > 1.0)
+  {    
+    if (globalInhibition_)
+    {
+      updateBoostFactorsGlobal_();
+    }
+    else
+    {
+      updateBoostFactorsLocal_();
+    }
+  }
+}
+
+void SpatialPooler::updateBoostFactorsGlobal_()
+{
+  Real targetDensity;
+  if (numActiveColumnsPerInhArea_ > 0)
   {
-    if (minActiveDutyCycles_[i] <= 0)
+    UInt inhibitionArea = pow((Real) (2 * inhibitionRadius_ + 1),
+                              (Real) columnDimensions_.size());
+    inhibitionArea = min(inhibitionArea, numColumns_);
+    targetDensity = ((Real) numActiveColumnsPerInhArea_) / inhibitionArea;
+    targetDensity = min(targetDensity, (Real) 0.5);
+  }
+  else
+  {
+    targetDensity = localAreaDensity_;
+  }
+
+  for (UInt i = 0; i < numColumns_; ++i)
+  {
+    Real boostFactor = exp(-(activeDutyCycles_[i] - targetDensity)
+                           * maxBoost_);
+
+    // Avoid floating point mismatches between implementations.
+    boostFactors_[i] = round(boostFactor * 100.0) / 100.0;
+  }
+ 
+}
+
+void SpatialPooler::updateBoostFactorsLocal_()
+{
+  vector<Real> targetDensity(numColumns_, 0);
+  
+  for (UInt i = 0; i < numColumns_; ++i)
+  {
+    UInt numNeighbors = 0;
+    Real localActivityDensity = 0;
+
+    if (wrapAround_)
     {
-      continue;
+      for (UInt neighbor : WrappingNeighborhood(i, inhibitionRadius_,
+                                                columnDimensions_))
+      {
+        localActivityDensity += activeDutyCycles_[neighbor];
+        numNeighbors += 1;
+      }      
     }
-    if (activeDutyCycles_[i] > minActiveDutyCycles_[i])
+    else
     {
-      boostFactors_[i] = 1.0;
-      continue;
+      for (UInt neighbor : Neighborhood(i, inhibitionRadius_,
+                                                columnDimensions_))
+      {
+        localActivityDensity += activeDutyCycles_[neighbor];
+        numNeighbors += 1;
+      }            
     }
-    boostFactors_[i] = ((1 - maxBoost_) / minActiveDutyCycles_[i] *
-                        activeDutyCycles_[i]) + maxBoost_;
+    targetDensity[i] = localActivityDensity / numNeighbors;
```

Comment:
Note that it makes sense to use the array of targetDensities in Python, since that allows us to do batch numpy operations rather than doing math in Python, which is slow. But in C we don't get any benefit from the vector (that I can see).